torch/_higher_order_ops/cond.py

```diff
@@ -371,7 +371,7 @@ def cond_op_dense(pred, true_fn, false_fn, operands):
     ), f"Dense implementation operands must be a list of tensors and ints {operands}"
     mode = _get_current_dispatch_mode()
     assert mode is None, "Mode should never be enabled for CPU/CUDA key"
-    if not torch.cuda.is_current_stream_capturing():
+    if not (torch.cuda.is_available() and torch.cuda.is_current_stream_capturing()):
```

Comment:
Since we're already making a custom TorchDispatchMode to do the warmup, could we put this logic in that class to avoid cudagraphs leaking here ?

Reply:
Good idea. I think I encountered some weirdness, that is described here: https://github.com/pytorch/pytorch/pull/140979/files#diff-4093d24068320f82997aecfaf4260aa495e3b4322e178cb5625f9555e3f8b0f5R388-R392

and here:

https://github.com/pytorch/pytorch/issues/140322

But let me take a second look. I would be happy to make that change, assuming there is nothing standing in the way.

Reply:
If https://github.com/pytorch/pytorch/issues/140322 is blocking we can do as a follow up

Reply:
I did what you suggested. Unfortunately we cannot use code within `__torch_dispatch__` right now to dispatch ops in a different way, but we can use py_impl(MyTorchDispatchMode) to do this right now, so I went with that workaround, though it is a little bit ugly because it causes some circular dependencies and separates code that should all sit in one single class. It should probably be redone once #140322 is handled. You can see how I do it for cond here: https://github.com/pytorch/pytorch/pull/140979/files#diff-4093d24068320f82997aecfaf4260aa495e3b4322e178cb5625f9555e3f8b0f5R384-R408

Note that I use two TorchDispatchModes, one for warmup, and one for making sure that I do stream capture appropriately for conditional nodes. I automatically enter CUDAGraphCaptureControlFlowOpDispatchMode in the __enter__ method of torch.cuda.graph. This unfortunately means that users who use CUDAGraph.capture_begin() and CUDAGraph.capture_end() instead of torch.cuda.graph() won't be able to capture conditional ops. I don't think this is a problem to be honest. Do you think it is?